src/ray/core_worker/object_recovery_manager.h

```diff
@@ -0,0 +1,151 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RAY_CORE_WORKER_OBJECT_RECOVERY_MANAGER_H
+#define RAY_CORE_WORKER_OBJECT_RECOVERY_MANAGER_H
+
+#include "absl/base/thread_annotations.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/common/id.h"
+#include "ray/core_worker/reference_count.h"
+#include "ray/core_worker/store_provider/memory_store/memory_store.h"
+#include "ray/core_worker/task_manager.h"
+#include "ray/raylet/raylet_client.h"
+
+namespace ray {
+
+typedef std::function<std::shared_ptr<PinObjectsInterface>(const std::string &ip_address,
+                                                           int port)>
+    ObjectPinningClientFactoryFn;
+
+typedef std::function<void(const ObjectID &object_id,
+                           const std::vector<rpc::Address> &raylet_locations)>
+    ObjectLookupCallback;
+
+class ObjectRecoveryManager {
```

Comment:
I assume this would only recover objects from common tasks, not actor tasks?

Reply:
Yes, it actually does not differentiate between actor and non-actor tasks, but by default actor tasks won't be reconstructed because they are added to the `TaskManager` with 0 maximum retries.